src/Ouroboros/Network/ChainFragment.hs

```diff
@@ -0,0 +1,561 @@
+    -- Recursively search the fingertree from the right
+    go Empty = False
+    go (c' :> b) | blockPoint b == p = True
+                 | otherwise         = go c'
+
+-- | Look for the intersection of the two 'ChainFragment's @c1@ and @c2@.
+--
+-- If they intersect, i.e., share a common 'Point', then return a tuple of:
+--
+-- * @l1@: the prefix of the first  chain fragment
+-- * @l2@: The prefix of the second chain fragment
+-- * @r1@: The suffix of the first  chain fragment
+-- * @r2@: The suffix of the second chain fragment
+--
+-- @l1@ and @l2@ will have same last block, i.e. /head/. The original chain
+-- @c1@ can be obtained by putting @r1@ after @l1@, similarly for @c2@: by
+-- putting @r2@ after @l2@.
+--
+-- Chains that intersect will always have the exact same common prefix, but
+-- chain fragments might have prefixes that differ in length. The blocks they
+-- have in common will be the same, but not each fragment might contain all
+-- the blocks the other fragment contains.
+--
+-- Take for example the following two chain fragments that share blocks 4 and
+-- 5. The two chain fragments are fragments of the same chain, but don't
+-- contain all blocks of the original chain. The missing history of the
+-- fragments is indicated with an asterisk (*). The @-A@ and @-B@ suffixes
+-- denote that blocks are part of a fork of the chain.
+--
+-- >
+-- >
+-- >     ┆ * ┆
+-- >     ├───┤
+-- >     │ 2 │     ┆ * ┆
+-- >     ├───┤     ├───┤
+-- >     │ 4 │     │ 4 │
+-- >     ├───┤     ├───┤
+-- >     │ 5 │     │ 5 │
+-- > ────┼───┼─────┼───┼───
+-- >     │ 6A│     │ 6B│
+-- >     └───┘     ├───┤
+-- >               │ 8B│
+-- >               └───┘
+-- >       c1        c2
+--
+-- The intersection of @c1@ and @c2@ is block 5 (the last 'Point' the two
+-- fragments have in common) and we return the following fragments:
+--
+-- >
+-- >
+-- >     ┆ * ┆
+-- >     ├───┤
+-- >     │ 2 │     ┆ * ┆
+-- >     ├───┤     ├───┤
+-- >     │ 4 │     │ 4 │
+-- >     ├───┤     ├───┤
+-- >     │ 5 │     │ 5 │
+-- > ────┴───┴─────┴───┴──────┬───┬─────┬───┬───
+-- >                          │ 6A│     │ 6B│
+-- >                          └───┘     ├───┤
+-- >                                    │ 8B│
+-- >                                    └───┘
+-- > Just (l1,       l2,        r1,       r2)
+intersectChainFragments
+  :: HasHeader block
+  => ChainFragment block -> ChainFragment block
+  -> Maybe (ChainFragment block, ChainFragment block,
+            ChainFragment block, ChainFragment block)
+intersectChainFragments initC1 initC2 = go initC1 initC2
+  where
+    go c1 c2 = case c2 of
+      Empty    -> Nothing
+      c2' :> b ->
+        let p = blockPoint b
+        in case splitAfterPoint c1 p of
+             Just (l1, r1)
+               | Just (l2, r2) <- splitAfterPoint initC2 p
+               -- splitAfterPoint initC2 p cannot fail, since p comes out of
+               -- initC2
+               -> Just (l1, l2, r1, r2)
+             _ -> go c1 c2'
```

Comment:
So just to check I've got it. This is O(|r2| * log |c1|) right? Since we're going down c2 block by block and looking for the point in the corresponding chain.

I'll have a think about whether we can do something that's log time in both sides.